packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/flat_dast/untagged_flat_dast.rs

```diff
@@ -0,0 +1,383 @@
+//! UntaggedFlatDast is part of the DAST normalization process where the DAST is flattened and all
+//! element/macro/etc. nodes are replaced with untagged references to their location in the nodes list.
+//! UntaggedFlatDast allows elements to change type without having to find all places where they are referenced.
+
+use serde::Serialize;
+
+use crate::dast::{DastElementContent, DastTextMacroContent};
+
+use super::super::{
+    macro_resolve::RefResolution, DastElement, DastError, DastFunctionMacro, DastMacro, DastRoot,
+    PathPart, Position,
+};
+
+pub use super::parent_iterator::ParentIterator;
+
+pub type Index = usize;
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(untagged)]
+pub enum UntaggedContent {
+    Text(String),
+    Ref(Index),
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct FlatElement {
+    pub name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub parent: Option<Index>,
+    pub children: Vec<UntaggedContent>,
+    pub attributes: Vec<FlatAttribute>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub position: Option<Position>,
+    pub idx: Index,
+    /// Information about the referent that this element extends (e.g., as specified by the `extend` attribute).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub extending: Option<RefResolution>,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct FlatAttribute {
+    pub name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub parent: Option<Index>,
```

Comment:
In what situation does a `FlatAttribute` not have a parent? Or is the `Option` there just to make processing easier by sharing the same format with other objects?

Reply:
I put it there to be like the others. If you're creating a the data structure outside of a tree, it is also nice to be able to put `None` in there instead of make up some value.